modules/ui/fields/date.js

```diff
@@ -0,0 +1,268 @@
+import { dispatch as d3_dispatch } from 'd3-dispatch';
+import { select as d3_select } from 'd3-selection';
+import * as countryCoder from '@ideditor/country-coder';
+
+import { uiCombobox } from '../combobox';
+import { t, localizer } from '../../core/localizer';
+import { utilGetSetValue, utilNoAuto, utilRebind, utilTotalExtent } from '../../util';
+
+
+export function uiFieldDate(field, context) {
+    let dispatch = d3_dispatch('change');
+    let yearInput = d3_select(null);
+    let eraInput = d3_select(null);
+    let monthInput = d3_select(null);
+    let dayInput = d3_select(null);
+    let _entityIDs = [];
+    let _tags;
+
+    let dateTimeFormat = new Intl.DateTimeFormat(localizer.languageCode(), {
+        year: 'numeric',
+        era: 'short',
+        month: 'long',
+        day: 'numeric',
+        timeZone: 'UTC',
+    });
+    let formatParts = dateTimeFormat.formatToParts(new Date()).map(p => p.type);
+
+    /**
+     * Returns the localized name of the era in the given format.
+     *
+     * @param year A representative year within the era.
+     */
+    function getEraName(year, format) {
+        let longFormat = new Intl.DateTimeFormat(localizer.languageCode(), {
+            year: 'numeric',
+            era: format,
+            timeZone: 'UTC',
+        });
+        let eraDate = new Date(Date.UTC(0, 0, 1));
+        eraDate.setUTCFullYear(year);
+        let parts = longFormat.formatToParts(eraDate);
+        let eraPart = parts.find(p => p.type === 'era');
+        return eraPart && eraPart.value;
+    }
+
+    let bceName = getEraName(0, 'short');
+    let ceName = getEraName(1, 'short');
```

Comment:
As far as I can tell, `Intl.DateTimeFormat` only exposes the locale’s default era names. Ideally we’d be using the secular era names, matching best practice in the relevant academic fields, but CLDR lets individual locales decide whether to provide secular names as the era names or variant era names, and ECMAScript doesn’t provide access to the variant era names. We might have to specify our own localizable strings for the two eras, even though many locales would have secular era names by default anyways.

Reply:
I opened tc39/ecma402#845 to request an addition to ECMAScript that would enable us to use the variant era names in CLDR. In the meantime, we’ll have to either live with the BC/AD names for now or maintain our own localizable strings. It’s no big deal to add more strings to core.yaml, but they won’t actually get translated because we haven’t set up a Translatewiki.net project for iD yet: OpenHistoricalMap/issues#470.